com.unity.multiplayer.mlapi/Runtime/Core/MessageBatcher.cs

```diff
@@ -0,0 +1,198 @@
+            length += messageStream.ReadByte() * 256;
+
+            return length;
+        }
+
+        /// <summary>
+        /// FillTargetList
+        /// Fills a list with the ClientId's an item is targeted to
+        /// </summary>
+        /// <param name="item">the FrameQueueItem we want targets for</param>
+        /// <param name="list">the list to fill</param>
+        private static void FillTargetList(in FrameQueueItem item, ref List<ulong> list)
+        {
+            list.Clear();
+            switch (item.QueueItemType)
+            {
+                case RPCQueueManager.QueueItemType.ServerRpc:
+                    list.Add(item.NetworkId);
+                    break;
+                case RPCQueueManager.QueueItemType.ClientRpc:
+                    foreach (ulong clientId in item.ClientIds)
+                    {
+                        list.Add(clientId);
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// QueueItem
+        /// Add a FrameQueueItem to be sent
+        /// </summary>queueItem
+        /// <param name="item">the threshold in bytes</param>
+        public void QueueItem(in FrameQueueItem item)
+        {
+            FillTargetList(item, ref TargetList);
+
+            foreach (ulong clientId in TargetList)
+            {
+                // todo: actually queue and buffer. For now, the dict contains just one entry !!!
+
+                if (!SendDict.ContainsKey(clientId))
+                {
+                    SendDict[clientId] = new SendStream();
+                    SendDict[clientId].Item = item;
+                    Writer.SetStream(SendDict[clientId].Stream);
+
+                    Writer.WriteBit(false); // Encrypted
+                    Writer.WriteBit(false); // Authenticated
+
+                    switch (item.QueueItemType)
+                    {
+                        // 6 bits are used for the message type, which is an MLAPIConstants
+                        case RPCQueueManager.QueueItemType.ServerRpc:
+                            Writer.WriteBits(MLAPIConstants.MLAPI_SERVER_RPC, 6); // MessageType
+                            break;
+                        case RPCQueueManager.QueueItemType.ClientRpc:
+                            Writer.WriteBits(MLAPIConstants.MLAPI_CLIENT_RPC, 6); // MessageType
+                            break;
+                    }
+                }
+
+                // write the amounts of bytes that are coming up
+                PushLength(item.MessageData.Count, ref Writer);
+
+                // write the message to send
+                // todo: is there a faster alternative to .ToArray()
+                Writer.WriteBytes(item.MessageData.ToArray(), item.MessageData.Count);
+
+                ProfilerStatManager.bytesSent.Record((int)item.MessageData.Count);
+                ProfilerStatManager.rpcsSent.Record();
```

Comment:
maybe we should put these under Editor and Debug macros so that they won't be in the standalone release builds?

Reply:
All the other stats don't seem to be under those macros. I went for "do as others are doing", to maintain consistency. It might be worthwhile, I don't know, but as a separate task and PR ?

Reply:
yeah... probably a separate task.

Reply:
Yeah, this sort of evolved from Scale Demo.  I like the idea of eliding it, but would be better if you didn't have to do ifdefs at all the callsites and instead codegen-it away / have null bodies